src/preload.py

```diff
@@ -609,8 +609,42 @@ def write_resin_device_pinning(app_data, output):
         )
 
 
+def isSupervisorApp(image, version):
+    images, supervisor_version, balena_os_version = (
+        get_images_and_supervisor_version()
+    )
+    if image in images and supervisor_version == version:
+        return True
+    return False
+
+
 def write_apps_json(data, output):
-    """Writes data dict to output as json"""
+    """Updates or creates preloaded state"""
+    if os.path.isfile(output):
+        with open(output, "r") as f:
+            current_state = json.load(f)
+        # Check state version consistency
+        for uuid in data.keys():
+            if len(uuid) != len(current_state.keys()[0]):
+                raise Exception("Current and target preloaded states not "
+                                "compatible")
+        # Remove all user apps from current state
+        for uuid in list(current_state['apps']):
+            app = current_state['apps'][uuid]
+            if 'isHost' in app:
+                continue
+            # Until supervisor is part of the hostOS app, there is no way to
+            # distinguish it from a single service user app from the state data
+            # The only way is checking the actual image contents
+            if len(app['services']) == 1:
+                for service in app['services']:
+                    image = app['services'][service]['image']
+                version = app['releaseVersion']
+                if isSupervisorApp(image, version) is True:
+                    continue
+            del current_state['apps'][uuid]
```

Comment:
> It looks like we can distinguish between host OS apps, supervisor apps and "user apps" 
We can cleanly distinguish user apps from hostOS app with the `isHost` attribute, which works once the supervisor becomes a hostOS app. Right now distinguishing the supervisor app from user apps is ugly - I need to read the image info and do a regex with the image name. Given that we know how this will be cleanly worked out in future I guess we can live with the hack.

>  when a user runs balena preload a second time for the same image file, preloading a different release the second time around, the contents of both releases are saved to the image file

That's correct. I have to say that after looking at this code it looks over-engineered and sub-optimal.

 > I would suggest that instead of deleting existing entries from apps.json (del current_state['apps'][uuid]), we threw an error 

That's a UX decision, I don't know how preloading is currently used. Are users version controlling a raw image and then updating the preloaded app as time goes by? Will asking them to always re-download the raw image be a problem? I am guessing not.

> Would balena preload be used to preload or replace host apps or supervisor apps in addition to "user apps"

Nothing has been discussed about this, and I expect not to be the case until we decide to expose hostOS apps to users which is not currently the case.

When that happens we can probably come up with a much cleaner implementation of the preloading engine that receives current state as input, prepares a data partition image to match, and writes this image over the data partition in the raw file. This could be the same transformation that happens at OS build time, and it could be reused in the cloud if required.

Reply:
I pushed the modification to exit when trying to re-preload.